### PR TITLE
feat(router): add v2 api for `/delete` `/add` `/retrieve`

### PR DESCRIPTION
**Pull request by ShankarSinghC**
add v2 api for `/delete` `/add` `/retrieve`

### CODE REVIEW

File: src/routes/data.rs [hunk 1]
```diff
@@ -31,6 +31,7 @@ use self::types::Validation;
 
 mod transformers;
 pub mod types;
+pub use super::routes_v2;
```
Comment (Chethan-rao) [inline src/routes/data.rs:34]:
```suggestion
use super::routes_v2;
```

File: src/routes/data.rs [hunk 2]
```diff
@@ -75,6 +76,15 @@ pub fn serve(
...
 
+    // v2 routes
+    let router = router.nest(
+        "/v2",
+        axum::Router::new()
+            .route("/vault/delete", post(routes_v2::data::delete_card))
+            .route("/vault/add", post(routes_v2::data::add_card))
+            .route("/vault/retrieve", post(routes_v2::data::retrieve_card)),
+    );
+
```
Comment (Chethan-rao) [inline src/routes/data.rs:87]: This will be under `/data/v2`. Is this fine?

CC: @NishantJoshi00

File: src/routes/routes_v2/data/types.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    pub entity_id: String,
+    pub vault_id: String,
+    pub status: Status,
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "PascalCase")]
+pub enum Status {
+    Ok,
+}
```
Comment (Chethan-rao) [inline src/routes/routes_v2/data/types.rs:20]: Do we need this Status? I think the HTTP status code itself is sufficient

File: src/storage/storage_v2.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+///
+/// VaultInterface:
+///
+/// Interface for interacting with the vault database table
+#[allow(dead_code)]
+pub(crate) trait VaultInterface {
+    type Algorithm: Encryption<Vec<u8>, Vec<u8>>;
+    type Error;
+
+    /// Fetch payment data from vault table by decrypting with `dek`
```
Comment (Chethan-rao) [inline src/storage/storage_v2.rs:17]: Please change the comment. Don't include any domain related things in v2

File: src/storage/storage_v2.rs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    async fn find_by_vault_id_entity_id(
+        &self,
+        vault_id: Secret<String>,
+        entity_id: &str,
+        key: &Self::Algorithm,
+    ) -> Result<types::Vault, ContainerError<Self::Error>>;
+
+    /// Insert payment data from vault table by decrypting with `dek`
```
Comment (Chethan-rao) [inline src/storage/storage_v2.rs:25]: Same here

File: src/storage/storage_v2.rs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    async fn insert_or_get_from_vault(
+        &self,
+        new: types::VaultNew,
+        key: &Self::Algorithm,
+    ) -> Result<types::Vault, ContainerError<Self::Error>>;
+
+    /// Delete card from the vault, without access to the `dek`
```
Comment (Chethan-rao) [inline src/storage/storage_v2.rs:32]: Same here

File: src/storage/storage_v2.rs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        &self,
+        vault_id: Secret<String>,
+        entity_id: &str,
+    ) -> Result<usize, ContainerError<Self::Error>>;
+
+    // async fn find_by_hash_id_merchant_id_customer_id(
+    //     &self,
+    //     entity_id: &str,
+    //     key: &Self::Algorithm,
+    // ) -> Result<Option<types::Vault>, ContainerError<Self::Error>>;
```
Comment (Chethan-rao) [inline src/storage/storage_v2.rs:43]: Can be removed?

File: src/storage/storage_v2/db.rs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        types::{StorageDecryption, StorageEncryption},
+        Storage,
+    },
+};
+
+use super::{types, VaultInterface};
+
+impl VaultInterface for Storage {
+    type Algorithm = GcmAes256;
+    type Error = error::LockerDBError;
```
Comment (Chethan-rao) [inline src/storage/storage_v2/db.rs:21]: Can we rename these Errors too as per v2

File: src/storage/storage_v2/types.rs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    pub entity_id: String,
+    pub enc_key: Secret<Vec<u8>>,
+    pub created_at: time::PrimitiveDateTime,
+}
+
+#[derive(Debug)]
+pub struct EntityNew<'a> {
+    pub entity_id: &'a str,
+    pub enc_key: Secret<Vec<u8>>,
+}
```
Comment (Chethan-rao) [inline src/storage/storage_v2/types.rs:23]: These are already defined in `v1` types. No need to define here again

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

File: src/app.rs [hunk 10]
```diff
@@ -108,6 +110,15 @@ where
     #[cfg(feature = "key_custodian")]
     let router = router.nest("/custodian", routes::key_custodian::serve());
 
+    // v2 routes
+    let router = router.nest(
+        "api/v2/vault",
```
Comment (Chethan-rao) [inline src/app.rs:115]:
```suggestion
        "/api/v2/vault",
```

File: src/storage/storage_v2.rs [hunk 11]
```diff
@@ -14,31 +14,25 @@ pub(crate) trait VaultInterface {
...
+    /// Fetch data from vault table
     async fn find_by_vault_id_entity_id(
         &self,
         vault_id: Secret<String>,
         entity_id: &str,
         key: &Self::Algorithm,
     ) -> Result<types::Vault, ContainerError<Self::Error>>;
 
-    /// Insert payment data from vault table by decrypting with `dek`
+    /// Insert data from vault table
```
Comment (Chethan-rao) [inline src/storage/storage_v2.rs:25]:
```suggestion
    /// Insert data into vault table
```

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by NishantJoshi00 — COMMENTED** [review]
This looks good, but i am not comfortable merging TODOs in the main branch, can we put them under v2 feature flag? Or at least let's wait for @Chethan-rao to raise a PR to this branch.

**Review by NishantJoshi00 — APPROVED** [review]

**Pull request merged**